src/platform/Darwin/DnssdInterfaceInfos.cpp

```diff
@@ -0,0 +1,183 @@
+    mInterfaceId          = if_nametoindex(ifa->ifa_name);
+    mInterfaceAddressType = ifa->ifa_addr->sa_family;
+
+    if (IsIPv4())
+    {
+        mInterfaceAddress.ipv4 = reinterpret_cast<struct sockaddr_in *>(ifa->ifa_addr)->sin_addr;
+    }
+    else
+    {
+        mInterfaceAddress.ipv6 = reinterpret_cast<struct sockaddr_in6 *>(ifa->ifa_addr)->sin6_addr;
+    }
+
+    return true;
+}
+
+DNSServiceErrorType InterfaceInfos::Register(DNSServiceRef sdRef, RegisterRecordContext * sdCtx, const char * hostname) const
+{
+    LogDetail();
+
+    uint16_t rrtype    = IsIPv4() ? kDNSServiceType_A : kDNSServiceType_AAAA;
+    uint16_t rdlen     = IsIPv4() ? 4 : 16;
+    const void * rdata = IsIPv4() ? static_cast<const void *>(GetIPv4()) : static_cast<const void *>(GetIPv6());
+
+    DNSRecordRef dnsRecordRef;
+    return DNSServiceRegisterRecord(sdRef, &dnsRecordRef, kRegisterRecordFlags, mInterfaceId, hostname, rrtype, kDNSServiceClass_IN,
+                                    rdlen, rdata, 0, OnRegisterRecord, sdCtx);
+}
+
+bool InterfaceInfos::HasValidFlags(unsigned int flags) const
+{
+    return !(flags & IFF_POINTOPOINT) && (flags & IFF_RUNNING) && (flags & IFF_MULTICAST);
+}
+
+bool InterfaceInfos::HasValidType(Inet::IPAddressType addressType, const sa_family_t addressFamily) const
+{
+    bool useIPv4 = addressType == chip::Inet::IPAddressType::kIPv4 || addressType == chip::Inet::IPAddressType::kAny;
+    bool useIPv6 = addressType == chip::Inet::IPAddressType::kIPv6 || addressType == chip::Inet::IPAddressType::kAny;
+
+    return (useIPv6 && addressFamily == AF_INET6) || (useIPv4 && addressFamily == AF_INET);
+}
+
+bool InterfaceInfos::IsValidInterfaceId(uint32_t targetInterfaceId, unsigned int currentInterfaceId) const
+{
+    return targetInterfaceId == kDNSServiceInterfaceIndexAny || targetInterfaceId == currentInterfaceId;
+}
+
+void InterfaceInfos::LogDetail() const
+{
+    char interfaceName[IFNAMSIZ] = {};
+    VerifyOrReturn(if_indextoname(mInterfaceId, interfaceName) != nullptr);
+
+    if (IsIPv4())
+    {
+        char addr[INET_ADDRSTRLEN] = {};
+        inet_ntop(AF_INET, GetIPv4(), addr, sizeof(addr));
+        ChipLogDetail(Discovery, "\tInterface: %s (%u) ipv4: %s", interfaceName, mInterfaceId, addr);
+    }
+    else
+    {
+        char addr[INET6_ADDRSTRLEN] = {};
+        inet_ntop(AF_INET6, GetIPv6(), addr, sizeof(addr));
+        ChipLogDetail(Discovery, "\tInterface: %s (%u) ipv6: %s", interfaceName, mInterfaceId, addr);
+    }
+}
+
+bool InterfacesInfos::Init(const char * hostname, chip::Inet::IPAddressType addressType, uint32_t interfaceId)
+{
+    mHostName = hostname;
+
+    // When the target interface is kDNSServiceInterfaceIndexLocalOnly, there is no need to map the hostname to
+    // an IP address.
+    VerifyOrReturnValue(interfaceId != kDNSServiceInterfaceIndexLocalOnly, true);
+
+    struct ifaddrs * ifap;
+    VerifyOrReturnValue(getifaddrs(&ifap) >= 0, false);
+
+    struct ifaddrs * ifa_start = ifap;
```

Comment:
Why is this needed?  It's not used for anything other than initializing the for loop start bit....